File: profile_filter.py
#!/usr/bin/env python3
"""
Profile-based content filtering for resume generation.

Filters resume content from content.yaml based on tag-matching profiles.
Used by both generate_latex_from_yaml.py and generate_docx_from_yaml.py
to produce role-targeted resumes from a single source of truth.

Filtering rules:
  - No profile selected → all content returned unchanged
  - Entry has no `tags` → always included
  - Entry has `tags` → included if ANY tag matches the profile's tag list
  - Achievement has no inline #hashtags → included when parent entry is included
  - Achievement has inline #hashtags → included if ANY hashtag matches profile tags
  - Limits applied after filtering (first N items kept, YAML order = priority)
  - All #hashtags stripped from output text
"""

import copy
import re
from typing import Optional

# Matches inline #hashtags like "#data" or "#community-ops"
_HASHTAG_RE = re.compile(r'\s*#([a-zA-Z][\w-]*)')


def strip_hashtags(text: str) -> str:
    """Remove inline #hashtag tokens from text."""
    return _HASHTAG_RE.sub('', text).rstrip()


def extract_hashtags(text: str) -> set[str]:
    """Extract the set of hashtag names from text (without the # prefix)."""
    return set(_HASHTAG_RE.findall(text))


def get_available_profiles(content: dict) -> list[dict]:
    """Return list of {name, label} for all defined profiles."""
    profiles = content.get('profiles', {})
    return [
        {'name': name, 'label': prof.get('label', name)}
        for name, prof in profiles.items()
    ]


def _tags_match(entry_tags: list | None, profile_tags: set) -> bool:
    """Check if an entry should be included based on its tags."""
    if not entry_tags:
        return True  # untagged = always included
    return bool(set(entry_tags) & profile_tags)


def _filter_achievements(achievements: list[str], profile_tags: set, max_achievements: int | None) -> list[str]:
    """Filter and strip achievements based on inline hashtags and limits."""
    filtered = []
    for achievement in achievements:
        hashtags = extract_hashtags(achievement)
        if not hashtags or hashtags & profile_tags:
            filtered.append(strip_hashtags(achievement))
    if max_achievements is not None:
        filtered = filtered[:max_achievements]
    return filtered


def _filter_entries(entries: list[dict], profile_tags: set, max_entries: int | None,
                    max_achievements: int | None = None) -> list[dict]:
    """Filter a list of entries (employment, education, projects) by tags and limits."""
    filtered = []
    for entry in entries:
        if not _tags_match(entry.get('tags'), profile_tags):
            continue
        entry = copy.deepcopy(entry)
        # Filter achievements if present
        if 'achievements' in entry and max_achievements is not None:
            entry['achievements'] = _filter_achievements(
                entry['achievements'], profile_tags, max_achievements
            )
        elif 'achievements' in entry:
            # Still need to strip hashtags even without a limit
            entry['achievements'] = [strip_hashtags(a) for a in entry['achievements']
                                     if not extract_hashtags(a) or extract_hashtags(a) & profile_tags]
        # Remove the tags key from output
        entry.pop('tags', None)
        filtered.append(entry)
    if max_entries is not None:
        filtered = filtered[:max_entries]
    return filtered


def _filter_skills(skills: dict, skills_tags: dict, profile_tags: set) -> dict:
    """Filter skills categories based on skills_tags mapping."""
    if not skills_tags:
        return skills
    filtered = {}
    for category, skill_list in skills.items():
        category_tags = skills_tags.get(category)
        if category_tags is None or set(category_tags) & profile_tags:
            filtered[category] = skill_list
    return filtered


def filter_content(content: dict, profile_name: Optional[str] = None) -> dict:
    """
    Apply profile-based filtering to resume content.

    Args:
        content: Full resume content dict (from YAML)
        profile_name: Name of the profile to apply, or None for no filtering

    Returns:
        Filtered content dict ready for generation
    """
    profiles = content.get('profiles', {})
    skills_tags = content.get('skills_tags', {})

    # Start with a shallow copy
    result = dict(content)

    # Always clean metadata keys from output
    result.pop('profiles', None)
    result.pop('skills_tags', None)

    # No profile → strip hashtags from achievements but return everything
    if not profile_name or profile_name not in profiles:
        # Still strip any hashtags from achievement text
        for section in ('employment', 'education', 'projects'):
            entries = result.get(section, [])
            cleaned = []
            for entry in entries:
                entry = copy.deepcopy(entry)
                if 'achievements' in entry:
                    entry['achievements'] = [strip_hashtags(a) for a in entry['achievements']]
                entry.pop('tags', None)
                cleaned.append(entry)
            result[section] = cleaned
        return result

    profile = profiles[profile_name]
    profile_tags = set(profile.get('tags', []))
    limits = profile.get('limits', {})

    # Replace summary if profile provides one
    if 'summary' in profile:
        result['summary'] = profile['summary']

    # Filter employment
    result['employment'] = _filter_entries(
        content.get('employment', []),
        profile_tags,
        limits.get('max_jobs'),
        limits.get('max_achievements'),
    )

    # Filter education
    result['education'] = _filter_entries(
        content.get('education', []),
        profile_tags,
        limits.get('max_education'),
    )

    # Filter projects
    result['projects'] = _filter_entries(
        content.get('projects', []),
        profile_tags,
        limits.get('max_projects'),
    )

    # Filter skills
    if 'skills' in content:
        result['skills'] = _filter_skills(content['skills'], skills_tags, profile_tags)

    # Set section order from profile
    if 'section_order' in profile:
        result['resume_modules'] = profile['section_order']

    return result

File: lib/profile-filter.ts
/**
 * Profile-based content filtering for resume display.
 *
 * Rules:
 *   - No profile → all content, achievements converted to plain strings
 *   - Untagged entries always included
 *   - Tagged entries included if ANY tag matches profile
 *   - Achievement tags filter at fine grain, priority sorts within filtered set
 *   - Limits truncate after filtering + sorting
 */

import type { ResumeData, AchievementInput, Achievement } from './resume';

// ── Legacy hashtag support (for plain string achievements) ──────────

const HASHTAG_RE = /\s*#([a-zA-Z][\w-]*)/g;

export function stripHashtags(text: string): string {
  return text.replace(HASHTAG_RE, '').trimEnd();
}

function extractHashtags(text: string): Set<string> {
  const tags = new Set<string>();
  let match;
  const re = /\s*#([a-zA-Z][\w-]*)/g;
  while ((match = re.exec(text)) !== null) {
    tags.add(match[1]);
  }
  return tags;
}

// ── Achievement helpers ─────────────────────────────────────────────

/** Normalize a mixed achievement to structured form */
function normalizeAchievement(a: AchievementInput): Achievement {
  if (typeof a === 'string') {
    const tags = Array.from(extractHashtags(a));
    return { text: stripHashtags(a), tags: tags.length > 0 ? tags : undefined };
  }
  return a;
}

/** Get tags from any achievement format */
function getAchievementTags(a: AchievementInput): string[] {
  if (typeof a === 'string') return Array.from(extractHashtags(a));
  return a.tags || [];
}

/** Get display text from any achievement format */
function getAchievementText(a: AchievementInput): string {
  if (typeof a === 'string') return stripHashtags(a);
  return a.text;
}

/** Get priority (undefined = lowest) */
function getAchievementPriority(a: AchievementInput): number | undefined {
  if (typeof a === 'string') return undefined;
  return a.priority;
}

/**
 * Filter achievements by profile tags, sort by priority, apply limit.
 * Returns plain strings for display.
 */
function filterAchievements(
  achievements: AchievementInput[],
  profileTags: Set<string>,
  maxAchievements?: number,
): string[] {
  // Filter: untagged = included, tagged = any match
  const filtered = achievements.filter((a) => {
    const tags = getAchievementTags(a);
    return tags.length === 0 || tags.some((t) => profileTags.has(t));
  });

  // Sort by priority: lower number first, undefined last, stable within same priority
  filtered.sort((a, b) => {
    const pa = getAchievementPriority(a);
    const pb = getAchievementPriority(b);
    if (pa === undefined && pb === undefined) return 0;
    if (pa === undefined) return 1;
    if (pb === undefined) return -1;
    return pa - pb;
  });

  // Extract text and apply limit
  const texts = filtered.map(getAchievementText);
  return maxAchievements != null ? texts.slice(0, maxAchievements) : texts;
}

/**
 * Convert all achievements to plain strings (no filtering, no sorting).
 * Used when no profile is selected.
 */
function flattenAchievements(achievements: AchievementInput[]): string[] {
  return achievements.map(getAchievementText);
}

// ── Shared helpers ──────────────────────────────────────────────────

function tagsMatch(entryTags: string[] | undefined, profileTags: Set<string>): boolean {
  if (!entryTags || entryTags.length === 0) return true;
  return entryTags.some((t) => profileTags.has(t));
}

/**
 * Resolve entry groups: if any entry in a group matched, pull in the rest of that group.
 * Entries with group_alt point to an alternative group — when the alt group is present,
 * this entry is excluded (they're mutually exclusive representations).
 */
function resolveGroups<T extends { group?: string; group_alt?: string }>(
  allEntries: T[],
  matchedEntries: T[],
): T[] {
  // Collect groups that matched
  const matchedGroups = new Set<string>();
  for (const e of matchedEntries) {
    if (e.group) matchedGroups.add(e.group);
  }

  // If no groups matched, nothing to resolve
  if (matchedGroups.size === 0) return matchedEntries;

  // Pull in unmatched entries whose group was matched
  const result = [...matchedEntries];
  const matchedSet = new Set(matchedEntries);
  for (const e of allEntries) {
    if (!matchedSet.has(e) && e.group && matchedGroups.has(e.group)) {
      result.push(e);
    }
  }

  // Remove entries whose group_alt is present (mutually exclusive)
  return result.filter((e) => {
    if (!e.group_alt) return true;
    return !matchedGroups.has(e.group_alt);
  });
}

export type ProfileInfo = {
  name: string;
  label: string;
};

export function getAvailableProfiles(content: ResumeData): ProfileInfo[] {
  const profiles = content.profiles ?? {};
  return Object.entries(profiles).map(([name, prof]) => ({
    name,
    label: prof.label ?? name,
  }));
}

// ── Normalize for editor use ────────────────────────────────────────

/** Convert all achievements in a ResumeData to structured Achievement objects */
export function normalizeAllAchievements(data: ResumeData): ResumeData {
  const result = { ...data };
  if (result.employment) {
    result.employment = result.employment.map((e) => ({
      ...e,
      achievements: e.achievements?.map(normalizeAchievement),
    }));
  }
  if (result.education) {
    result.education = result.education.map((e) => ({
      ...e,
      achievements: e.achievements?.map(normalizeAchievement),
    }));
  }
  return result;
}

// ── Filter by raw tags (AI matching) ────────────────────────────────

export function filterByTags(content: ResumeData, tags: string[]): ResumeData {
  const profileTags = new Set(tags);
  const result: ResumeData = { ...content };
  delete result.profiles;
  delete result.skills_tags;

  if (content.employment) {
    const matched = content.employment.filter((e) => tagsMatch(e.tags, profileTags));
    const resolved = resolveGroups(content.employment, matched);
    result.employment = resolved.map((e) => {
      const cleaned = { ...e };
      delete cleaned.tags;
      delete cleaned.group;
      delete cleaned.group_alt;
      if (cleaned.achievements) {
        cleaned.achievements = filterAchievements(cleaned.achievements, profileTags);
      }
      return cleaned;
    });
  }

  if (content.education) {
    const matched = content.education.filter((e) => tagsMatch(e.tags, profileTags));
    const resolved = resolveGroups(content.education, matched);
    result.education = resolved.map((e) => {
      const cleaned = { ...e };
      delete cleaned.tags;
      delete cleaned.group;
      delete cleaned.group_alt;
      if (cleaned.achievements) {
        cleaned.achievements = filterAchievements(cleaned.achievements, profileTags);
      }
      return cleaned;
    });
  }

  if (content.projects) {
    result.projects = content.projects
      .filter((p) => tagsMatch(p.tags, profileTags))
      .map((p) => {
        const cleaned = { ...p };
        delete cleaned.tags;
        return cleaned;
      });
  }

  const skillsTags = content.skills_tags ?? {};
  if (content.skills && Object.keys(skillsTags).length > 0) {
    const filteredSkills: Record<string, string[]> = {};
    for (const [category, skillList] of Object.entries(content.skills)) {
      const catTags = skillsTags[category];
      if (!catTags || catTags.some((t: string) => profileTags.has(t))) {
        filteredSkills[category] = skillList;
      }
    }
    result.skills = filteredSkills;
  }

  return result;
}

// ── Filter by named profile ─────────────────────────────────────────

export function filterContent(content: ResumeData, profileName?: string): ResumeData {
  const profiles = content.profiles ?? {};
  const skillsTags = content.skills_tags ?? {};

  const result: ResumeData = { ...content };
  delete result.profiles;
  delete result.skills_tags;

  // No profile → flatten achievements to strings, remove tags, return everything
  // Show detailed group entries, hide summary alternates (group_alt entries)
  if (!profileName || !(profileName in profiles)) {
    for (const section of ['employment', 'education'] as const) {
      const entries = result[section];
      if (!entries) continue;
      (result as any)[section] = entries
        .filter((entry: any) => !entry.group_alt)
        .map((entry: any) => {
          const cleaned = { ...entry };
          delete cleaned.tags;
          delete cleaned.group;
          delete cleaned.group_alt;
          if (cleaned.achievements) {
            cleaned.achievements = flattenAchievements(cleaned.achievements);
          }
          return cleaned;
        });
    }
    if (result.projects) {
      result.projects = result.projects.map((p) => {
        const cleaned = { ...p };
        delete cleaned.tags;
        return cleaned;
      });
    }
    return result;
  }

  const profile = profiles[profileName];
  const profileTags = new Set(profile.tags ?? []);
  const limits = profile.limits ?? {};

  if (profile.summary) {
    result.summary = profile.summary;
  }

  // Filter employment
  if (content.employment) {
    const matched = content.employment.filter((e) => tagsMatch(e.tags, profileTags));
    const resolved = resolveGroups(content.employment, matched);
    let filtered = resolved.map((e) => {
      const cleaned = { ...e };
      delete cleaned.tags;
      delete cleaned.group;
      delete cleaned.group_alt;
      if (cleaned.achievements) {
        cleaned.achievements = filterAchievements(
          cleaned.achievements,
          profileTags,
          limits.max_achievements,
        );
      }
      return cleaned;
    });
    if (limits.max_jobs != null) filtered = filtered.slice(0, limits.max_jobs);
    result.employment = filtered;
  }

  // Filter education
  if (content.education) {
    const matched = content.education.filter((e) => tagsMatch(e.tags, profileTags));
    const resolved = resolveGroups(content.education, matched);
    let filtered = resolved.map((e) => {
      const cleaned = { ...e };
      delete cleaned.tags;
      delete cleaned.group;
      delete cleaned.group_alt;
      if (cleaned.achievements) {
        cleaned.achievements = filterAchievements(cleaned.achievements, profileTags);
      }
      return cleaned;
    });
    if (limits.max_education != null) filtered = filtered.slice(0, limits.max_education);
    result.education = filtered;
  }

  // Filter projects
  if (content.projects) {
    let filtered = content.projects
      .filter((p) => tagsMatch(p.tags, profileTags))
      .map((p) => {
        const cleaned = { ...p };
        delete cleaned.tags;
        return cleaned;
      });
    if (limits.max_projects != null) filtered = filtered.slice(0, limits.max_projects);
    result.projects = filtered;
  }

  // Filter skills
  if (content.skills && Object.keys(skillsTags).length > 0) {
    const filteredSkills: Record<string, string[]> = {};
    for (const [category, skillList] of Object.entries(content.skills)) {
      const catTags = skillsTags[category];
      if (!catTags || catTags.some((t: string) => profileTags.has(t))) {
        filteredSkills[category] = skillList;
      }
    }
    result.skills = filteredSkills;
  }

  // Set section order
  if (profile.section_order) {
    const order = profile.section_order.filter((m: string) => m !== 'header');
    result.resume_modules = ['header', ...order];
  }

  return result;
}
